[mlir][bufferization][NFC] Rename copy_tensor op to materialize_in_destination

matthias-springer opened the pull request:
The previous name was badly chosen. The op is used to ensure that a computation materializes in the future buffer of a certain tensor.

Depends On #65766. Only review the top commit.

maerhart commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationOps.td:
```diff
@@ -209,17 +209,30 @@ def Bufferization_CloneOp : Bufferization_Op<"clone", [
+
+    Note: "tensor.insert_slice" could be used for the same purpose, but since
+    tensor dialect ops only indicate *what* should be computed but not *where*,
+    it could fold away, causing the computation to materialize in a different
+    buffer.
```
If this lowers to a memcpy doesn't it also materialize in a different buffer? Do you maybe have a concrete example in mind that you could add here? Or explain in a bit more detail why materializing in a different buffer is a problem?

matthias-springer replied:
It could lower to something like `memref.memcpy %x, %x`. That's the case when the source and the destination tensor are equivalent. E.g.:
```
%0 = arith.select %c, %t, %t
%1 = bufferization.materialize_in_destination %0 into %t
```
In the above example that's easy to see and it could fold away, but it may not obvious in more complex cases (e.g., with tiling, nested loops, etc.).

maerhart replied:
Thanks! That helped understanding a lot. Should we add it to the docs as well?

maerhart reviewed (commented):
Do I understand correctly that this is not just renaming but also relaxing the guarantees the op provides (guaranteed to lower to a memcpy vs only lowers to a memcpy if needed)?
I assume there are no users of this op that relied on the guarantee of it to lower to a memcpy?

matthias-springer committed:
[mlir][bufferization][NFC] Rename copy_tensor op to materialize_in_destination

The previous name was badly chosen. The op is used to ensure that a computation materializes in the future buffer of a certain tensor.
